### PR TITLE
[feat] Register chains and tokens through precompile

### LINKED ISSUE #115: Non-determinism in GetClientChains function due to map iteration
The  function in  retrieves all client chain IDs but iterates over a map to do so, which could introduce non-determinism in the output order of chain IDs. This issue needs to be addressed to ensure consistent behavior, especially if the order of IDs is significant. This will be handled in a future PR.

### PR DESCRIPTION
**Pull request by TimmyExogenous**
## Description

Closes #115

This PR implements tokens and client chain registration functions through the precompiled contract. 
Additionally, it fix issue#115 regarding the non-determinism caused by using a map.

## Summary by CodeRabbit

- **New Features**
  - Added settings for `exocore_lz_app_address` and `meta_info` under assets, and included a new token "Tether USD."

- **Improvements**
  - Enhanced parameter clarity in deposit, withdraw, client chain, and token functions.
  - Introduced pagination for querying all client chain info.
  - Improved error messages for better clarity.
  
- **Bug Fixes**
  - Fixed naming consistency from `CheckExocoreLzAppAddr` to `CheckExocoreGatewayAddr`.
  - Corrected error messages for address length validation.

- **Refactor**
  - Renamed methods for better clarity and consistency.
  - Updated function names to reflect improved functionality.

### CODE REVIEW

File: local_node.sh
Comment (MaxMustermann2) [inline local_node.sh:89]: Is this the CREATE2 address?

File: precompiles/assets/types.go
Comment (MaxMustermann2) [inline precompiles/assets/types.go:93]: Validate length of metaInfo non-zero ? We can look at setting a maximum size for the lengths later

File: precompiles/assets/types.go
Comment (MaxMustermann2) [inline precompiles/assets/types.go:102]: Same here, should we validate non-empty?

File: precompiles/assets/types.go
Comment (MaxMustermann2) [inline precompiles/assets/types.go]: Add TODO for non-EVM chains

File: x/assets/keeper/client_chain.go
Comment (MaxMustermann2) [inline x/assets/keeper/client_chain.go:45]: For my understanding, this is the part that makes it deterministic right? The iteration order is sorted in the store.

**Review by MaxMustermann2 — APPROVED** [review]
Some questions, but otherwise looks good.

File: precompiles/assets/IAssets.sol
Comment (MaxMustermann2) [inline precompiles/assets/IAssets.sol:54]: These should be added to `IsTransaction`

File: precompiles/reward/parser.go
Comment (MaxMustermann2) [reply precompiles/reward/parser.go]: `stakerAddr` not `WithdrawRewardAddress`

**Review by MaxMustermann2 — APPROVED** [review]

File: local_node.sh
Comment (TimmyExogenous) [reply local_node.sh:89]: No, it's the default address from the above local mnemonic. It is used to start the local node, then directly use this EOA address to test all precompiled contract interfaces, without relying on the client chain and Exocore gateway contract.

File: x/assets/keeper/client_chain.go
Comment (TimmyExogenous) [reply x/assets/keeper/client_chain.go:45]: Yeah

File: precompiles/assets/IAssets.sol
Comment (TimmyExogenous) [reply precompiles/assets/IAssets.sol:54]: Yeah, thanks

File: precompiles/assets/types.go
Comment (TimmyExogenous) [reply precompiles/assets/types.go]: It appears that the compatibility of non-EVM chains will be addressed in the Exocore gateway contract. However, it's still pending, so I will add a code comment.

File: precompiles/assets/types.go
Comment (TimmyExogenous) [reply precompiles/assets/types.go:93]: The initial version had a validation for this. I removed it to keep it consistent because it isn't validated in the genesis validation and this field had been set in local_node.sh. It's better to set a maximum size. I can add a maximum and the validation in both the precompile and the genesis validation.

File: precompiles/assets/types.go
Comment (TimmyExogenous) [reply precompiles/assets/types.go:102]: This field can be left empty, and there will be a default signature type when the input is empty. However, the final decision will be made after researching the signature algorithms of most blockchains.

File: precompiles/assets/types.go
Comment (TimmyExogenous) [reply precompiles/assets/types.go]: Thanks, I will remove the usage of reflect.

File: precompiles/assets/assets_test.go
Comment (TimmyExogenous) [reply precompiles/assets/assets_test.go:62]: Maybe just renaming `TestRunWithdrawThroughClientChain` to `TestRunWithdrawPrincipal` is fine.

File: precompiles/assets/types.go
Comment (TimmyExogenous) [reply precompiles/assets/types.go]: I will avoid the usage of reflect

File: x/assets/keeper/params.go
Comment (TimmyExogenous) [reply x/assets/keeper/params.go:42]: Seem like the error has been handled in `GetExocoreGatewayAddress`

File: precompiles/common/error.go
Comment (TimmyExogenous) [reply precompiles/common/error.go]: I just want to use an Error to address the different cases of invalid address length.

File: precompiles/slash/parser.go
Comment (TimmyExogenous) [reply precompiles/slash/parser.go]: I just want to use an Error to address the different cases of invalid address length."

File: precompiles/assets/assets.go
Comment (TimmyExogenous) [reply precompiles/assets/assets.go]: Thanks, but the returned should be `method.Outputs.Pack(false)`

File: precompiles/slash/parser.go
Comment (TimmyExogenous) [reply precompiles/slash/parser.go]: The codes regarding the slash and reward module should be deprecated or re-implemented.

File: precompiles/reward/parser.go
Comment (TimmyExogenous) [reply precompiles/reward/parser.go]: The codes regarding the slash and reward module should be deprecated or re-implemented.

File: precompiles/reward/parser.go
Comment (TimmyExogenous) [reply precompiles/reward/parser.go]: The codes regarding the slash and reward module should be deprecated or re-implemented.

File: precompiles/delegation/delegation.go
Comment (TimmyExogenous) [reply precompiles/delegation/delegation.go]: Thanks

File: x/assets/keeper/grpc_query.go
Comment (TimmyExogenous) [reply x/assets/keeper/grpc_query.go]: thanks

File: precompiles/assets/types.go
Comment (TimmyExogenous) [reply precompiles/assets/types.go:93]: Additionally, generate.js might need to be changed slightly after this PR is merged.

**Comment by cloud8little** [comment]
Test passed. 

1. Register client chain through precompile contract.
```
exocored q assets QueAllClientChainInfo
all_client_chain_infos:
- address_length: 20
  chain_id: "0"
  exocore_chain_index: "0"
  finalization_blocks: "0"
  layer_zero_chain_id: "101"
  meta_info: Example EVM chain meta info
  name: Example EVM chain
  signature_type: ""
- address_length: 20
  chain_id: "0"
  exocore_chain_index: "0"
  finalization_blocks: "0"
  layer_zero_chain_id: "40159"
  meta_info: BSC
  name: Binance Smart Chain
  signature_type: ECDSA
pagination:
  next_key: null
  total: "2"
```

2. Register new asset through precompile

```
exocored q assets QueAllStakingAssetsInfo
all_staking_assets_info:
  0x1e867667ef16111047c2c7f6adf4612bdf80064d_0x9cdf:
    asset_basic_info:
      address: 0x1e867667ef16111047c2c7f6adf4612bdf80064d
      decimals: 18
      exocore_chain_index: "0"
      layer_zero_chain_id: "40159"
      meta_info: ""
      name: ""
      symbol: ""
      total_supply: "2000000000000000000"
    staking_total_amount: "0"
```

**Review by cloud8little — APPROVED** [review]

File: x/assets/keeper/params.go
Comment (bwhour) [inline x/assets/keeper/params.go:47]: can we rename this **ExocoreLzAppAddress** to ExocoreGatewayAddress which is easy to understand? 

Additionally, If in the future **exocoreBtcGateway** and **exocoreSolGateway**  etc. cannot be integrated with **exocoregateway**, how should this part of the logic be handled?

File: local_node.sh
Comment (MaxMustermann2) [reply local_node.sh:89]: Ah okay - it is the `local_funded_account`. Can you please add a comment for future reference?

File: precompiles/assets/types.go
Comment (MaxMustermann2) [reply precompiles/assets/types.go:93]: > It's better to set a maximum size. I can add a maximum and the validation in both the precompile and the genesis validation.

Can you create an issue for tracking this?

File: local_node.sh
Comment (TimmyExogenous) [reply local_node.sh:89]: Yeah

File: x/assets/keeper/params.go
Comment (TimmyExogenous) [reply x/assets/keeper/params.go:47]: Yeah, it should be renamed to ExocoreGatewayAddress. However, the renaming might cause inconsistency in the data within the KVStore, so we need to address these issues together at an appropriate time. I have created an issue to document these concerns
https://github.com/ExocoreNetwork/exocore/issues/113

Comment (TimmyExogenous) [reply x/assets/keeper/params.go:47]: > Additionally, If in the future exocoreBtcGateway and exocoreSolGateway etc. cannot be integrated with exocoregateway, how should this part of the logic be handled?

This depends on the design for Bitcoin and Solana. If we use a unified gateway address to address all client chains, then it doesn't need to be changed. We might store a list for the gateway address if we use multiple gateway contracts.

File: precompiles/assets/types.go
Comment (TimmyExogenous) [reply precompiles/assets/types.go:107]: Yeah, this error has provided clear and actionable information to the caller.

File: precompiles/assets/types.go
Comment (TimmyExogenous) [reply precompiles/assets/types.go]: It doesn't seem necessary to make such modifications.

File: precompiles/assets/types.go
Comment (TimmyExogenous) [reply precompiles/assets/types.go:93]: I have added the maximum and validation in the last commit.

File: precompiles/assets/tx.go
Comment (adu-web3) [inline precompiles/assets/tx.go]:
```suggestion
func (p Precompile) RegisterToken(
```

File: precompiles/assets/types.go
Comment (adu-web3) [reply precompiles/assets/types.go:93]: I wonder if we could allow this `metaInfo` being empty? since it is extra data that is used for arbitrary data like an url or plain text. I don't check this field to be non-empty in the contract implementation.

File: precompiles/assets/types.go
Comment (adu-web3) [inline precompiles/assets/types.go:101]: should we use an enum value instead of string if we could exhaust existed common signature types. Of course we could also leave it for future PRs if we haven't researched into this

File: precompiles/assets/types.go
Comment (adu-web3) [inline precompiles/assets/types.go:120]: I see `p.assetsKeeper.GetClientChainInfoByIndex` would return error if client chain is not registered before, just comment to get your confirmation about this if my understanding is ture

File: precompiles/assets/types.go
Comment (adu-web3) [inline precompiles/assets/types.go:141]: In `ExocoreGateway`, I check that tvlLimit is not zero because zero tvl limit seems not make sense. what do you think?

File: precompiles/assets/types.go
Comment (adu-web3) [inline precompiles/assets/types.go:126]: in`ExocoreGateway` I check that the asset address is non-zero, do you think it is needed here?

File: precompiles/assets/types.go
Comment (adu-web3) [inline precompiles/assets/types.go:160]: same as above comment it seems reasonable to allow metaInfo being empty

**Review by adu-web3 — APPROVED** [review]
overall look good to me, I have some comments mainly about parameters validation for precompile functions `registerClientChain` and `registerToken`

File: precompiles/assets/types.go
Comment (TimmyExogenous) [reply precompiles/assets/types.go:93]: The `metaInfo` isn't allowed to be empty, the check was added in the last commit. It seems like the check should be implemented in the front end to avoid invalid transactions. As for the contract, would It be enough to check it in the precompile implementation?

File: precompiles/assets/tx.go
Comment (TimmyExogenous) [reply precompiles/assets/tx.go]: thanks

File: precompiles/assets/types.go
Comment (TimmyExogenous) [reply precompiles/assets/types.go:101]: Using the string type might be clearer when querying or setting values. However, using enum can reduce gas costs. This can be decided later when this field is actually used.

File: precompiles/assets/types.go
Comment (TimmyExogenous) [reply precompiles/assets/types.go:120]: Yeah, the asset registration will depend on the related client chain's registration.

File: precompiles/assets/types.go
Comment (TimmyExogenous) [reply precompiles/assets/types.go:141]: Yeah, I will add a check for this.

File: precompiles/assets/types.go
Comment (TimmyExogenous) [reply precompiles/assets/types.go:126]: The length of assetAddr will be checked to ensure it is not smaller than 20, so it is fine whether or not it is checked in the `ExocoreGateway`.

File: precompiles/assets/types.go
Comment (TimmyExogenous) [reply precompiles/assets/types.go:160]: It seems better not to allow the metaInfo to be empty if we provide this field to describe the client chain. This can ensure detailed and clearer information about the client chain for all users. Of course, this is a minor issue that won't have a significant impact no matter how it's handled.

File: precompiles/assets/tx.go
Comment (TimmyExogenous) [reply precompiles/assets/tx.go:34]: It seems not necessary.

File: precompiles/assets/types.go
Comment (adu-web3) [reply precompiles/assets/types.go:160]: ok, I'd better add this check in contract as well

File: precompiles/assets/types.go
Comment (adu-web3) [reply precompiles/assets/types.go:126]: this is different: even if the `assetAddr` is `address(0)`, the length of `assetAddr` would be 20

File: precompiles/assets/types.go
Comment (adu-web3) [reply precompiles/assets/types.go:93]: From design perspective, I think your question actually means: whether native module should totally trust in `ExocoreGateway` or native module should do defensive parameter validation regardless the implementation of `ExocoreGateway`? personally I prefer the native module not totally trusting the implementation of `ExocoreGateway` and check for params even if they might have been checked in `ExocoreGateway`

**Review by bwhour — APPROVED** [review]

File: precompiles/assets/types.go
Comment (TimmyExogenous) [reply precompiles/assets/types.go:93]: Yeah, the params will be checked in the native module no matter how it's implemented in `ExocoreGateway`.

File: precompiles/assets/types.go
Comment (TimmyExogenous) [reply precompiles/assets/types.go:126]: I see your point. It doesn't seem necessary to check these special addresses here. If we are checking address(0), what about other special addresses like address(1) or address(0xffff)? It might be better to include these checks when registering the assets, even if we intend to address them. I'm not sure if checking these special addresses during DApp development is a default specification. If it is, adding related address checks during registration might be more standard.

Comment (adu-web3) [reply precompiles/assets/types.go:126]: ok, we can skip this as not neccessary

**Pull request merged**
